java/src/main/java/ai/rapids/cudf/Table.java

```diff
@@ -1044,6 +1080,136 @@ public Table aggregate(Aggregate... aggregates) {
         aggregate.close();
       }
     }
+
+    public Table aggregateWindows(WindowAggregate... windowAggregates) {
+      // To improve performance and memory we want to remove duplicate operations
+      // and also group the operations by column so hopefully cudf can do multiple aggregations
+      // in a single pass.
+
+      // Use a tree map to make debugging simpler (columns are all in the same order)
+      TreeMap<Integer, ColumnWindowOps> groupedOps = new TreeMap<>(); // Map agg-col-id -> Agg ColOp.
```

Comment:
nit: it looks like there is a lot of overlap between the rolling window functions and the general aggregate function. It would be good to have a follow on issue to see if there is any way to make the code more common.  It is really minor so I would prefer that we don't try to get that done now and add more risk to this.

Reply:
I agree. In fact, I did try move commonality out to a separate function template. My memory is that my first attempt wasn't easy.

After this PR is merged, I'll raise a separate JIRA for this on GitHub to attempt to boil this down.